File: app/src/main/java/net/veldor/flibustaloader/utils/BookSharer.java
package net.veldor.flibustaloader.utils;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.os.StrictMode;
import android.provider.DocumentsContract;
import android.util.Log;
import android.widget.Toast;

import androidx.documentfile.provider.DocumentFile;

import net.veldor.flibustaloader.App;
import net.veldor.flibustaloader.R;

import java.io.File;

import static android.content.Intent.FLAG_ACTIVITY_NEW_TASK;
import static android.content.Intent.FLAG_GRANT_READ_URI_PERMISSION;
import static net.veldor.flibustaloader.utils.BookOpener.intentCanBeHandled;

public class BookSharer {

    public static void shareBook(String name, String type) {
        Log.d("surprise", "BookSharer shareBook " + type);
        File file;
        // ========================================================================================
        Context context = App.getInstance();
        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            DocumentFile downloadsDir = App.getInstance().getDownloadDir();
            if (downloadsDir != null) {
                DocumentFile downloadFile = downloadsDir.findFile(name);
                if (downloadFile != null) {
                    final String docId;
                    docId = DocumentsContract.getDocumentId(downloadFile.getUri());
                    Log.d("surprise", "BookSharer shareBook " + docId);
                    final String[] split = docId.split(":");
                    final String storage = split[0];
                    String path = "///storage/" + storage + "/" + split[1];
                    file = new File(path);
                    Log.d("surprise", "BookSharer shareBook " + file);
                    // костыли, проверю существование файла с условием, что он находится на основной флешке
                    if (!file.exists()) {
                        file = new File(Environment.getExternalStorageDirectory() + "/" + split[1]);
                    }
                    if (file.exists()) {
                        //todo По возможности- разобраться и заменить на валидное решение
                        StrictMode.VmPolicy.Builder builder = new StrictMode.VmPolicy.Builder();
                        StrictMode.setVmPolicy(builder.build());
                        // отправлю запрос на открытие файла
                        Intent shareIntent = new Intent(Intent.ACTION_SEND);
                        shareIntent.putExtra(Intent.EXTRA_STREAM, Uri.fromFile(file));
                        shareIntent.setType(MimeTypes.getFullMime(type));
                        shareIntent.addFlags(FLAG_GRANT_READ_URI_PERMISSION | FLAG_ACTIVITY_NEW_TASK);
                        if (intentCanBeHandled(shareIntent)) {
                            App.getInstance().startActivity(shareIntent);
                        } else {
                            Toast.makeText(App.getInstance(), "Не найдено приложение, открывающее данный файл", Toast.LENGTH_SHORT).show();
                        }
                    } else {
                        Toast.makeText(context, context.getString(R.string.file_not_found_message), Toast.LENGTH_LONG).show();
                    }
                } else {
                    Toast.makeText(context, context.getString(R.string.file_not_found_message), Toast.LENGTH_LONG).show();
                }
            }
        } else {
            File dd = MyPreferences.getInstance().getDownloadDir();
            File bookFile = new File(dd, name);
            if (bookFile.isFile()) {
                //todo По возможности- разобраться и заменить на валидное решение
                StrictMode.VmPolicy.Builder builder = new StrictMode.VmPolicy.Builder();
                StrictMode.setVmPolicy(builder.build());
                // отправлю запрос на открытие файла
                Intent shareIntent = new Intent(Intent.ACTION_SEND);
                shareIntent.putExtra(Intent.EXTRA_STREAM, Uri.fromFile(bookFile));
                shareIntent.setType(MimeTypes.getFullMime(type));
                shareIntent.addFlags(FLAG_GRANT_READ_URI_PERMISSION | FLAG_ACTIVITY_NEW_TASK);
                if (intentCanBeHandled(shareIntent)) {
                    App.getInstance().startActivity(shareIntent);
                } else {
                    Toast.makeText(App.getInstance(), "Не найдено приложение, открывающее данный файл", Toast.LENGTH_SHORT).show();
                }
            } else {
                Toast.makeText(context, context.getString(R.string.file_not_found_message), Toast.LENGTH_LONG).show();
            }
        }
    }

    public static void shareLink(String link) {
        Context context = App.getInstance();
        Intent i = new Intent(Intent.ACTION_SEND);
        i.setType("text/plain");
        i.putExtra(Intent.EXTRA_SUBJECT, R.string.share_url_message);
        i.putExtra(Intent.EXTRA_TEXT, link);
        Intent starter = Intent.createChooser(i, context.getString(R.string.share_url_title));
        starter.addFlags(FLAG_ACTIVITY_NEW_TASK);
        if (TransportUtils.intentCanBeHandled(starter)) {
            context.startActivity(starter);
        } else {
            Toast.makeText(context, "Упс, не нашлось приложения, которое могло бы это сделать.", Toast.LENGTH_LONG).show();
        }
    }
}

File: app/src/main/java/net/veldor/flibustaloader/utils/BookSharer.kt
package net.veldor.flibustaloader.utils

import android.content.Context
import android.os.Build
import net.veldor.flibustaloader.App
import android.content.Intent
import android.widget.Toast
import net.veldor.flibustaloader.R
import android.provider.DocumentsContract
import android.os.Environment
import android.os.StrictMode.VmPolicy
import android.os.StrictMode
import android.net.Uri
import android.util.Log
import java.io.File

object BookSharer {


    fun shareBook(name: String) {
        // find the book
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            val downloadsDir = PreferencesHandler.instance.getDownloadDir()
            if(downloadsDir != null){
                val targetFile = FilesHandler.find(downloadsDir.listFiles(), name)
                if(targetFile != null){
                    // найден файл, делюсь им
                    val docId: String = DocumentsContract.getDocumentId(targetFile.uri)
                    val split = docId.split(":").toTypedArray()
                    val storage = split[0]
                    val path = "///storage/" + storage + "/" + split[1]
                    var file = File(path)
                    // костыли, проверю существование файла с условием, что он находится на основной флешке
                    if (!file.exists()) {
                        file = File(
                            Environment.getExternalStorageDirectory().toString() + "/" + split[1]
                        )
                    }
                    if (file.exists()) {
                        //todo По возможности- разобраться и заменить на валидное решение
                        val builder = VmPolicy.Builder()
                        StrictMode.setVmPolicy(builder.build())
                        // отправлю запрос на открытие файла
                        val shareIntent = Intent(Intent.ACTION_SEND)
                        shareIntent.putExtra(Intent.EXTRA_STREAM, Uri.fromFile(file))
                        shareIntent.type = MimeTypes.getFullMime(Grammar.getExtension(name))
                        shareIntent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_ACTIVITY_NEW_TASK)
                        App.instance.startActivity(
                            Intent.createChooser(
                                shareIntent,
                                App.instance.getString(R.string.send_book_title)
                            ).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                        )
                    } else {
                        Toast.makeText(
                            App.instance,
                            App.instance.getString(R.string.file_not_found_message),
                            Toast.LENGTH_LONG
                        ).show()
                    }
                }
            }
            else{
                Toast.makeText(
                    App.instance,
                    "Не распознана папка загрузок, выберите ещё ещё раз в настройках",
                    Toast.LENGTH_SHORT
                ).show()
            }
        }
        else{
            val dd: File = PreferencesHandler.instance.compatDownloadDir!!
            dd.walk().filter { it.name == name }
        }
    }

    @kotlin.jvm.JvmStatic
    fun shareLink(link: String?) {
        val context: Context = App.instance
        val i = Intent(Intent.ACTION_SEND)
        i.type = "text/plain"
        i.putExtra(Intent.EXTRA_SUBJECT, R.string.share_url_message)
        i.putExtra(Intent.EXTRA_TEXT, link)
        val starter = Intent.createChooser(i, context.getString(R.string.share_url_title))
        starter.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        if (TransportUtils.intentCanBeHandled(starter)) {
            context.startActivity(starter)
        } else {
            Toast.makeText(
                context,
                "Упс, не нашлось приложения, которое могло бы это сделать.",
                Toast.LENGTH_LONG
            ).show()
        }
    }
}